charts/DiscreteBarTransform.ts

```diff
@@ -168,15 +186,21 @@ export class DiscreteBarTransform implements IChartTransform {
         })
 
         const data = sortBy(allData, d => d.value)
-
         const colorScheme = chart.baseColorScheme
             ? ColorSchemes[chart.baseColorScheme]
             : undefined
-        const colors = colorScheme ? colorScheme.getColors(data.length) : []
+        const uniqValues = uniq(data.map(d => d.value))
+        const colors = colorScheme?.getColors(uniqValues.length) || []
         if (chart.props.invertColorScheme) colors.reverse()
 
-        data.forEach((d, i) => {
-            d.color = chart.data.keyColors[d.key] || colors[i] || d.color
+        const colorByValue = new Map<number, string>()
+        uniqValues.forEach((value, i) => colorByValue.set(value, colors[i]))
+
+        data.forEach(d => {
+            d.color =
+                chart.data.keyColors[d.key] ||
+                colorByValue.get(d.value) ||
+                d.color
```

Comment:
I'm not quite sure where `allData` is used, and whether we need this logic in there at all?

Reply:
Yeah our code here is not very clean and there is some duplication of logic. We probably need to extract some common logic from `allData` and `currentData` in one place. I think the colouring logic is not needed here (though I might be wrong).

Basically, there can be a timeline on bar charts, which is not implemented in the "DiscreteBar" chart type, but you can see the timeline if you are looking at a LineChart and you collapse the timeline to a single year: https://ourworldindata.org/grapher/annual-co2-emissions-per-country?tab=chart&time=1908

While you are dragging the timeline on a bar chart, in order to avoid scale changes, we need to find the maximum value for all possible years, so that we can calculate the space needed and keep the layout consistent. So we iterate through `allData` to get all years, not just `currentData` which contains data for the currently selected year.

(I think this is at least what I remember, might not be exactly correct.)

Reply:
Ah, I see. I was just confused on whether these timelines were in use at all.

Just being curious: What's the difference between DiscreteBarChart and LineChart, then?

Reply:
Maybe not a direct answer, but: The long-term intent is to allow users to navigate across all charts that are useful for the data they are looking at.

Showing the bar chart when they collapse the timeline of the line chart is one step in that direction. Another we are planning to make is to allow readers to switch between line chart and slope chart – they are very similar charts, a slope chart is basically a line chart where only the values for start and end years are shown, and the values in the middle are removed.